src/charm.py

```diff
@@ -674,7 +691,16 @@ def _generate_grafana_config(self) -> str:
         can be set in ENV variables, but leave for expansion later so we can
         hide auth secrets
         """
-        return self._generate_database_config() if self.has_db else ""
+        db_config = self._generate_database_config() if self.has_db else ""
```

Comment:
this reads a bit odd.
I'd refactor to use some if/elses:
```suggestion
cfgs = []
if self.has_db:
   cfgs.append(self._generate_db_config())
if self.has_cert:
    cfgs.append(self._generate_network_config())
return "\n".join(cfgs)
```